src/merkle_tree.rs

```diff
@@ -217,8 +216,32 @@ where
     }
 
     pub fn build_merkle_proof(leaves: &[T], leaf_indices: &[u32]) -> Option<MerkleProof<T, M>> {
-        // TODO: Remove this clone
-        Self::build_merkle_tree(leaves.to_vec()).build_proof(leaf_indices)
+        Self::build_merkle_tree(leaves).build_proof(leaf_indices)
+    }
+
+    /// retrieve that a proof points to leaves of a tree, returning `None` if the proof indices is empty or out of bounds
+    pub fn retrieve_leaves(leaves: &[T], proof: &MerkleProof<T, M>) -> Option<Vec<T>> {
```

Comment:
The requirement `&[T]` is too strong. The minimal requirements set is:

* Know the count of leaves.
* Get the value of a leaf at the specific index.

The slice requires all the leaves values are present, but we are only interested in several leaves.

For example, in CKB, we don't need to read all transactions to verify the proof. We can construct the transaction key from index and only read transactions on demand.

Reply:
Leave this as a future task when we really need such kind of performance tuning.